packages/transaction-metrics-exporter/src/block-processor.ts

```diff
@@ -0,0 +1,266 @@
+import { newLogExplorer } from '@celo/contractkit/lib/explorer/log-explorer'
+import { Histogram, labelValues, linearBuckets } from 'prom-client'
+import { Transaction } from 'web3-core'
+import { Block } from 'web3-eth'
+
+import { Counters } from './metrics'
+import { Contracts, stateGetters } from './states'
+import { toMethodId, toTxMap } from './utils'
+
+import { tracerAsText } from './tracer'
+
+enum LoggingCategory {
+  Block = 'RECEIVED_BLOCK',
+  ParsedLog = 'RECEIVED_PARSED_LOG',
+  ParsedTransaction = 'RECEIVED_PARSED_TRANSACTION',
+  State = 'RECEIVED_STATE',
+  Transaction = 'RECEIVED_TRANSACTION',
+  TransactionReceipt = 'RECEIVED_TRANSACTION_RECEIPT',
+  InternalTransaction = 'INTERNAL_TRANSACTION',
+}
+
+interface DataResult {
+  contract: string
+  function: string
+  args: any
+  blockNumber: number
+  values: { [metric: string]: number }
+}
+
+export class BlockProcessor {
+  private contracts: Contracts = {} as any
+  private initialized = false
+  private histograms: { [key: string]: Histogram } = {}
+
+  constructor(
+    private kit: ContractKit,
+    private blockInterval = 1,
+    private fromBlock: number = 0,
+    private toBlock?: number
+  ) {}
+
+  async init() {
+    if (this.initialized) {
+      throw new Error('BlockProcessor is running')
+    }
+    this.initialized = true
+
+    await this.loadContracts()
+
+    if (this.fromBlock?.toFixed?.()) {
+      await this.initBatch()
+      if (!this.toBlock) {
+        await this.initSubscription()
+      }
+    } else {
+      await this.initSubscription()
+    }
+  }
+
+  async loadContracts() {
+    this.contracts.Exchange = await this.kit.contracts.getExchange()
+    this.contracts.SortedOracles = await this.kit.contracts.getSortedOracles()
+    this.contracts.Reserve = await this.kit.contracts.getReserve()
+    this.contracts.GoldToken = await this.kit.contracts.getGoldToken()
+    this.contracts.EpochRewards = await this.kit.contracts.getEpochRewards()
+    this.contracts.StableToken = await this.kit.contracts.getStableToken()
+  }
+
+  async initSubscription() {
+    const subscription = await this.kit.web3.eth.subscribe('newBlockHeaders')
+
+    // Prevent same block multiples times
+    let lastBlocks: number[] = []
+    subscription.on('data', (header) => {
+      if (!lastBlocks.includes(header.number)) {
+        // tslint:disable-next-line: no-floating-promises
+        this.onNewBlock(header.number)
+      }
+      lastBlocks.push(header.number)
+      lastBlocks = lastBlocks.slice(-10)
+    })
+  }
+
+  async initBatch() {
+    let block = this.fromBlock - 1
+    while (++block <= (await this.getToBlock())) {
+      await this.onNewBlock(block, block - this.fromBlock)
+    }
+  }
+
+  async getToBlock() {
+    if (this.toBlock) {
+      return this.toBlock
+    }
+    return (await this.kit.web3.eth.getBlock('latest')).number
+  }
+
+  async onNewBlock(blockNumber: number, blockIntervalNumber?: number) {
```

Comment:
What's the usecase for not just processing every single block?

Reply:
For doing quick sampling long range for analysis (stability, reserve, rewards, exchange, ...)